lib/internal/readline/interface.js

```diff
@@ -1178,14 +1203,38 @@ class Interface extends InterfaceConstructor {
     if (this[kIsMultiline] && rows > 0) {
       const splitLine = StringPrototypeSplit(this.line, '\n');
       const previousLine = splitLine[rows - 1];
+      let amountToMove = 0;
+      // Move to the beginning of the current line + 1 char to go to the end of the previous line
+      const amountToClamp = -cols + 1;
+
       // If I am moving up and the current line is longer than the previous line
-      const amountToMove = (cols > previousLine.length + 1) ?
-        -cols + 1 : // Move to the beginning of the current line + 1 char to go to the end of the previous line
-        -previousLine.length - 1; // Otherwise just move to the previous line, in the same position
+      const shouldClampColumn = (cols > previousLine.length + 1);
+      if (shouldClampColumn) {
+        if (this[kPreviousCursorCols] === -1) {
+          this[kPreviousCursorCols] = cols; // Save the cursor cols position to restore it later
+        }
+        amountToMove = amountToClamp;
+      } else {
+        // When pressing up a couple of times passing through shorter lines.
+        // Move to the previous line, in the same position
+        amountToMove = -previousLine.length - 1;
+        if (this[kPreviousCursorCols] !== -1) {
+          if (this[kPreviousCursorCols] <= previousLine.length) {
+            // If, while moving up, the cursor passed through a line shorter than where the cursor was
+            // I need to move it back to the original position
+            amountToMove += this[kPreviousCursorCols] - cols;
+            this[kResetPreviousCursorCols](); // Reset it to avoid moving again
+          } else {
+            amountToMove = amountToClamp; // Go to the end of the previous line
+          }
+        }
```

Comment:
That code seems to be a almost 1-to-1 duplication and could be abstracted. I think that's worth doing due to the number of lines duplicated.

Reply:
yeah you are right, just pushed a more unified function :)